### PR TITLE
Documentation Update to 0.4.0

### PR DESCRIPTION
**Pull request by soundeffects**
The documentation has not only been updated to 0.4.0, but has also been largely rewritten in a way I believe brings more clarity and approachability, especially in tutorials.

There are a few changes I'm still working on before this pull request is ready, including:
- Change of Simphony package version in the config to 0.4.0
- Finishing the "Developing" and "Documenting" pages in the documentation source
- Fixing SiPANN code documentation so that Sphinx auto-documenting features work for it
- Fixing the warnings Sphinx is currently encountering when auto-documenting SiEPIC

I'd appreciate feedback on the writing style and level of detail in the tutorials.

### CODE REVIEW

**Comment by soundeffects** [comment]
Warnings are still being raised by Sphinx when auto-documenting SiEPIC, but I don't believe they are causing any problems. Fixing SiPANN auto-docs will require the Github Actions that build the docs to have SiPANN installed. It may not be worth the hassle to auto-doc SiPANN models. I've also realised that the Github README needs updating.

These are all features that should be looked at in future, but I believe that this pull request will be fine to merge without them.

File: docs/source/reference/api.rst [hunk 1]
```diff
@@ -1,40 +1,44 @@
...
 
 .. automodule:: simphony
-    :members:
-    :inherited-members:
-    :show-inheritance:
+  :members:
+  :inherited-members:
+  :show-inheritance:
 
 .. automodule:: simphony.connect
```
Comment (AustP) [inline docs/source/reference/api.rst]: `simphony.connect` is an internal module only. We don't really want end-users to know about this module / use it in any way. So let's not include it in the docs.

File: docs/source/reference/libraries.rst [hunk 2]
```diff
@@ -0,0 +1,10 @@
+.. _model-libraries:
+
+Model Libraries
+===============
+
+.. automodule:: simphony.libraries.siepic
+  :members:
+
+.. automodule:: simphony.libraries.sipann
+  :members:
```
Comment (AustP) [inline docs/source/reference/libraries.rst:10]: The images for the SiEPIC library are broken.
Also, It looks like `simphony/libraries/sipann.py` is missing the pydoc string for the module.

**Review by AustP — CHANGES_REQUESTED** [review]
The docs look great! The new organization is excellent.
I made a few requests, which shouldn't take too long and then we'll get v0.4.0 pushed.

File: docs/source/reference/libraries.rst (hunk 2 above)
Comment (soundeffects) [reply docs/source/reference/libraries.rst:10]: Not sure I get what you mean by "images are broken". I have added a pydoc string at the beginning of the SiPANN library, though.

File: docs/source/reference/api.rst (hunk 1 above)
Comment (soundeffects) [reply docs/source/reference/api.rst]: Just removed it.

File: docs/source/reference/libraries.rst (hunk 2 above)
Comment (soundeffects) [reply docs/source/reference/libraries.rst:10]: Ok, I had removed the images and that's why they aren't working. I just put the images back in, and I also created a new image for Halfrings, though it doesn't match the style of the other diagrams.

**Review by AustP — APPROVED** [review]

**Pull request merged**